geoopt/optim/rwolfe.py

```diff
@@ -0,0 +1,689 @@
+            have changed in between consecutive steps. (default: False)
+        no_step : bool (optional)
+            If True, just compute step size and do not perform the step.
+            (default: False)
+        """
+
+        self.closure = closure
+        phi0, derphi0 = self._init_loss(recompute_gradients=recompute_gradients)
+        self._step_size_dic = dict()
+
+        phi_information = {
+            "phi0": phi0,
+            "derphi0": derphi0,
+            "old_phi0": self.old_phi0,
+            "old_derphi0": self.old_derphi0,
+            "old_step_size": self.last_step_size,
+        }
+        step_size = self.line_search_method(
+            self._phi, self._derphi, **phi_information, **self.line_search_params
+        )
+
+        self.step_size_history.append(step_size)
+        self.last_step_size = step_size
+
+        self.old_phi0 = phi0
+        self.old_derphi0 = derphi0
+
+        # Ensure that the last step for which we computed the closure coincides with
+        # proposed step size, so that we can reuse the gradients and retract.
+        # This is very rare, and should only happen if force_step=True and no stepsize
+        # was found, or for user-defined linesearch methods.
+        if self._last_step != step_size or not self.compute_derphi:
+            self._grads_computed = False
+        redo_retract = self._last_step != step_size
+
+        # If it fails to find a good step, probably we have convergence
+        if step_size is None:
+            if force_step:
+                step_size = self.fallback_stepsize
+                self._grads_computed = False
+            elif (
+                self.last_step_size is None
+            ):  # Warn if step_size is None twice in a row
+                warning_string = """No suitable step size could be found, and no step
+                was taken. Call `step` with `force_step=True` to take a step anyway.
+                """
+                warnings.warn(warning_string, LineSearchWarning)
+
+        for point in self._params:
+            if step_size is None or no_step:
+                continue
+            state = self.state[point]
+            if "search_direction" not in state:
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = Euclidean()
+
+            search_direction = state["search_direction"]
+
+            # Compute retract if suggested step size is not the last one we tried (rare)
+            if redo_retract:
+                new_point = manifold.retr(point, step_size * search_direction)
+            else:
+                new_point = state["new_point"]
+
+            # Use retract to perform the step, and transport the search direction
+            if self.transport_search_direction:
+                search_transported = manifold.transp_follow_retr(
```

Comment:
Using `transp_follow_retr` here seems less efficient in general than using `transp`.  However, for CanonicalStiefel for example, `transp` is not defined but `transp_follow_retr` is.